### ISSUE TITLE
make signing agnostic of key tweaking

### ISSUE THREAD

**Issue by matthiasgeihs** [jesseposner/FROST-BIP340#14]
i found that the signing protocol can be made agnostic to key tweaking.
basically, one can prepare the key share and public key ahead of time, and then run the same protocol as without tweaking.

i think this would simplify the code a bit. if there is interest, i can give it a shot and create a pr with the relevant changes. wanted to check here first, if this is desired / what others think.

**Comment by jonasnick** [comment]
Hi @matthiasgeihs, can you go a little bit more into detail how that'd look like and what the trade-offs are API-wise?

**Comment by matthiasgeihs** [comment]
> Hi @matthiasgeihs, can you go a little bit more into detail how that'd look like and what the trade-offs are API-wise?
> 
> EDIT: you may want to bring up the idea in the [FROST-signing draft BIP](https://github.com/siv2r/bip-frost-signing)

see here: https://github.com/jesseposner/FROST-BIP340/compare/main...matthiasgeihs:FROST-BIP340:feat/pre-tweak

this shows that the aggregator can be made agnostic of key tweaking (if i'm not mistaken)

**Comment by jonasnick** [comment]
Thanks for the example. I think I get it. So the idea is that you tweak the shares such that the sum of the partial signatures _is_ the s value of the final signature. Therefore, partial signature aggregation does not need to be aware of the tweaks. Sounds reasonable.

We could have done the same in the MuSig2 spec, but I think we didn't because we wanted the aggregator to be able to identify faulty signers. In our thinking so far, that includes that the aggregator has the ability to check that the tweaks have been applied correctly using partial signature verification on the tweaked keys.

**Comment by jonasnick** [comment]
> I am not very familiar with MuSig2 and how signatures shares are validated there.

MuSig2 signing is very similar to FROST. The only real difference is that instead of multiplying the secret key with a lagrange coefficient you multiply with a key aggregation coefficient.

> you can basically apply the same pre-tweak to the commitment shares and then use these to validate the tweaked signature shares

I'm not sure exactly what you mean by commitment shares. I can imagine that what we call the "pubshares" (the share the signer uses for signing times the generator) can be already tweaked by the signers. Perhaps that's what you're saying. I think that's fine in FROST. But it would be possible that the aggregator has long-term untweaked pubshares and the signers apply their individual tweaks for each signing session, or that the aggregator even provides the tweaks to the signer. Arguably that stretches the "aggregator" role quite a bit.

**Comment by matthiasgeihs** [comment]
yep, sorry, i was using the wrong term. i meant to refer to the "public verification shares". see my edit on my post above.

in our case, we don't use the verification shares, but the general idea is that the signing parties provide a service to the authenticated user. the authenticated user asks the key share holders to sign messages on request. the user can be seen in a similar role as an aggregator, who is the recipient of the final signature, but who also instructs the shareholders to use a key tweak in the signing protocol.

the benefit of integrating key tweaking like i've shown in my PR was that we were able to keep the core protocol library as it is, and only update the code of the services that use this library (by tweaking the inputs, like the key share, before running the core library). so basically the core library is agnostic of whether key tweaking is used or not. for me, this was a worthwhile observation.

**Comment by jonasnick** [comment]
> this was a worthwhile observation

Agree. I think we should consider this for the FROST API.

**Comment by siv2r** [comment]
> So the idea is that you tweak the shares such that the sum of the partial signatures is the s value of the final signature

Oh, neat trick! I noticed this being used in [ChillDKG’s `invalid_taproot_commit`](https://github.com/BlockstreamResearch/bip-frost-dkg/blob/678088911e867094b113f0d28bc3f3782cc975c3/python/chilldkg_ref/simplpedpop.py#L209-L211) method for tweaking the threshold public key. I briefly discussed this with @jesseposner about potentially applying it to BIP Signing but didn’t actively pursue it. Thanks for bringing it up @matthiasgeihs, I'll open an issue to track this for BIP Signing.

> We could have done the same in the MuSig2 spec

I don't think this approach will work with MuSig2. AFAIU, it works in FROST due to its Shamir shares: if you offset the secret shares by a constant $k$, the shared secret also shifts by $k$, [because of Lagrange interpolation](https://x.com/siv2r/status/1870487837541646601). In MuSig2, additive shares don’t behave the same way, and there’s also additional randomization involved.
